src/tests/algorithms/test_generic.py

```diff
@@ -40,3 +40,169 @@ def test_rescale_image(self):
         self.assertSequenceEqual(s, (4, 1.0, 0.25, 1.0))
 
 
+class TestBoxcarFilter(unittest.TestCase):
+
+    _kernel_size_0D: tuple = ()
+    _kernel_size_1D = (3,)
+    _kernel_size_2D = (3, 3)
+    _kernel_size_3D = (3, 3, 5)
+
+    _signal_1D = [1.25, 4.52, 46.36]
+    _signal_2D = [[1.25, 4.52, 46.36],
+                 [3.45, 25.66, 6234.6]]
+
+    _known_inputs_outputs_1D = [
+        ([],
+         []),
+
+        ([5.],
+         [5.]),
+
+        (np.array([4., 6.]),
+         np.array([14., 16.]) / 3.),
+
+        (np.ones(5),
+         np.ones(5)),
+
+        (np.array([1., 2., 1.]),
+         np.array([4., 4., 4.]) / 3.),
+
+        (np.array([0., 1., 2., 3.]),
+         np.array([1., 3., 6., 8.]) / 3.),
+
+        (np.array([-1,  1., 2., 3., -3.]),
+         np.array([-1., 2., 6., 2., -3.]) / 3.),
+
+        (np.array([-1,   1., -1., 1., -1., 1.]),
+         np.array([-1., -1.,  1, -1.,  1., 1.]) / 3.),
+
+        (np.array([-0.5, 1.5, 2.5, 3.5, -2.5]),
+         np.array([ 0.5, 3.5, 7.5, 3.5, -1.5]) / 3.),
+    ]
+
+    _known_inputs_outputs_2D = [
+        ([[],
+          []],
+         [[],
+          []]),
+
+        ([[5.],
+          [5.]],
+         [[5.],
+          [5.]]),
+
+        (np.array([[4., 6.],
+                   [4., 6.]]),
+         np.array([[14., 16.],
+                   [14., 16.]]) / 3.),
+
+        (np.ones((5, 5)),
+         np.ones((5, 5))),
+
+        (np.array([[1., 2., 1.],
+                   [1., 2., 1.]]),
+         np.array([[4., 4., 4.],
+                   [4., 4., 4.]]) / 3.),
+
+        (np.array([[0., 1., 2., 3.],
+                   [0., 1., 2., 3.]]),
+         np.array([[1., 3., 6., 8.],
+                   [1., 3., 6., 8.]]) / 3.),
+
+        (np.array([[-1,  1., 2., 3., -3.],
+                   [-1,  1., 2., 3., -3.]]),
+         np.array([[-1., 2., 6., 2., -3.],
+                   [-1., 2., 6., 2., -3.]]) / 3.),
+
+        (np.array([[-1,   1., -1., 1., -1., 1.],
+                  [-1,   1., -1., 1., -1., 1.]]),
+         np.array([[-1., -1.,  1, -1.,  1., 1.],
+                   [-1., -1.,  1, -1.,  1., 1.]]) / 3.),
+
+        (np.array([[-0.5, 1.5, 2.5, 3.5, -2.5],
+                   [-0.5, 1.5, 2.5, 3.5, -2.5]]),
+         np.array([[ 0.5, 3.5, 7.5, 3.5, -1.5],
+                   [ 0.5, 3.5, 7.5, 3.5, -1.5]]) / 3.),
+    ]
+
+
+    def assertFilterOutput(self, signal, kernel, out_exp):
+        """Assert that signal generates output close to expectation"""
+        with self.subTest(signal=signal):
+            out_fnd = boxcar_filter(signal, kernel)
+            msg = ''
+            try:
+                np.testing.assert_allclose(out_fnd, out_exp)
+            except AssertionError as e:
+                msg = '\nsignal: %s\ngives wrong filtered output:' % (signal,)
+                msg += str(e).replace('x:', '  actual:').replace('y:', 'expected:')
+
+            if msg:
```

Comment:
You can remove the `if msg:` by moving the statement into the `except` part.

Reply:
That is true from the program logic point of view and that is what it was initially. However, it turns out that if you do that, you get a warning that an exception was raised (by self.fail()) inside an exception handler. It does not seem to like that.